File: src/pct/history.cc
// Copyright 2011 The University of Michigan
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Authors - Jie Yu (jieyu@umich.edu)

// File: pct/history.cc - Implementation of the PCT scheduler history.

#include "pct/history.h"

#include <fstream>

namespace pct {

unsigned long History::AvgInstCount() {
  double total = 0.0;
  int size = table_proto_.history_size();
  for (int i = 0; i < size; i++) {
    total += (double)table_proto_.history(i).inst_count();
  }
  return (unsigned long)(total / (double)size);
}

unsigned long History::AvgNumThreads() {
  double total = 0.0;
  int size = table_proto_.history_size();
  for (int i = 0; i < size; i++) {
    total += (double)table_proto_.history(i).num_threads();
  }
  return (unsigned long)(total / (double)size);
}

void History::Update(unsigned long inst_count, unsigned long num_threads) {
  HistoryProto *proto = table_proto_.add_history();
  proto->set_inst_count(inst_count);
  proto->set_num_threads(num_threads);
}

void History::Load(const std::string &file_name) {
  std::fstream in(file_name.c_str(), std::ios::in | std::ios::binary);
  table_proto_.ParseFromIstream(&in);
  in.close();
}

void History::Save(const std::string &file_name) {
  std::fstream out(file_name.c_str(),
                   std::ios::out | std::ios::trunc | std::ios::binary);
  table_proto_.SerializeToOstream(&out);
  out.close();
}

} // namespace pct

File: script/maple/pct/history.py
"""Copyright 2011 The University of Michigan

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

Authors - Jie Yu (jieyu@umich.edu)
"""

import os
from maple.core import proto

def history_pb2():
    return proto.module('pct.history_pb2')

class HistoryEntry(object):
    def __init__(self, proto, db):
        self.proto = proto
        self.db = db
    def inst_count(self):
        return self.proto.inst_count
    def num_threads(self):
        return self.proto.num_threads
    def __str__(self):
        content = []
        content.append('%-4d' % self.num_threads())
        content.append('%d' % self.inst_count())
        return ' '.join(content)

class History(object):
    def __init__(self):
        self.proto = history_pb2().HistoryTableProto()
        self.history = []
    def load(self, db_name):
        if not os.path.exists(db_name):
            return
        f = open(db_name, 'rb')
        self.proto.ParseFromString(f.read())
        f.close()
        for history_proto in self.proto.history:
            entry = HistoryEntry(history_proto, self)
            self.history.append(entry)
    def avg_inst_count(self):
        total = 0.0
        for entry in self.history:
            total += entry.inst_count()
        return total / len(self.history)
    def avg_num_threads(self):
        total = 0.0
        for entry in self.history:
            total += entry.num_threads()
        return total / len(self.history)
    def display(self, f):
        for entry in self.history:
            f.write('%s\n' % str(entry))
    def display_summary(self, f):
        f.write('PCT History Summary\n')
        f.write('---------------------------\n')
        f.write('Avg inst count   = %f\n' % self.avg_inst_count())
        f.write('Avg num threads  = %f\n' % self.avg_num_threads())

